File: examples/02-gpio-interrupts/gpio-interrupts.c
/*
 * Advanced GPIO Example - Interrupt-style Event Detection
 * 
 * This example shows how to use libgpiod's event detection features,
 * which are similar to interrupts in embedded systems.
 * 
 * Features demonstrated:
 * - Edge detection (rising/falling edges)
 * - Event-driven programming (no polling)
 * - Timeouts and non-blocking I/O
 * - Multiple GPIO handling
 * 
 * Hardware setup:
 * - LED on GPIO17 (pin 11)
 * - Button on GPIO27 (pin 13) 
 * - Optional: Second button on GPIO22 (pin 15)
 */

#include <gpiod.h>
#include <stdio.h>
#include <unistd.h>
#include <signal.h>
#include <stdbool.h>
#include <string.h>     // For memset
#include <sys/time.h>   // For timeval structure
#include <errno.h>      // For error codes

#define GPIO_CHIP "gpiochip4"
#define LED_PIN 17
#define BUTTON1_PIN 27
#define BUTTON2_PIN 22

// Global running flag for signal handler
static volatile bool running = true;

// Structure to hold our GPIO context
// Structures in C group related data together
typedef struct {
    struct gpiod_chip *chip;
    struct gpiod_line *led;
    struct gpiod_line *button1;
    struct gpiod_line *button2;
    int button1_count;  // Count button presses
    int button2_count;
} gpio_context_t;

void signal_handler(int sig) {
    printf("\nReceived signal %d, shutting down...\n", sig);
    running = false;
}

/*
 * Initialize GPIO hardware
 * This function demonstrates proper error handling with cleanup
 * 
 * @param ctx: Pointer to GPIO context structure
 * @return: 0 on success, -1 on failure
 */
int gpio_init(gpio_context_t *ctx) {
    // Clear the structure (good practice)
    memset(ctx, 0, sizeof(gpio_context_t));
    
    // Open GPIO chip
    ctx->chip = gpiod_chip_open_by_name(GPIO_CHIP);
    if (!ctx->chip) {
        fprintf(stderr, "Failed to open %s: %s\n", GPIO_CHIP, strerror(errno));
        return -1;
    }
    
    // Get GPIO lines
    ctx->led = gpiod_chip_get_line(ctx->chip, LED_PIN);
    ctx->button1 = gpiod_chip_get_line(ctx->chip, BUTTON1_PIN);
    ctx->button2 = gpiod_chip_get_line(ctx->chip, BUTTON2_PIN);
    
    if (!ctx->led || !ctx->button1 || !ctx->button2) {
        fprintf(stderr, "Failed to get GPIO lines\n");
        return -1;
    }
    
    // Configure LED as output
    if (gpiod_line_request_output(ctx->led, "gpio-interrupts", 0) < 0) {
        fprintf(stderr, "Failed to request LED output\n");
        return -1;
    }
    
    // Configure buttons for event detection
    // FALLING_EDGE means detect HIGH->LOW transition (button press with pull-up)
    // This is similar to configuring interrupt triggers in embedded systems
    if (gpiod_line_request_falling_edge_events_flags(ctx->button1, "gpio-interrupts",
            GPIOD_LINE_REQUEST_FLAG_BIAS_PULL_UP) < 0) {
        fprintf(stderr, "Failed to request button1 events\n");
        return -1;
    }
    
    if (gpiod_line_request_falling_edge_events_flags(ctx->button2, "gpio-interrupts",
            GPIOD_LINE_REQUEST_FLAG_BIAS_PULL_UP) < 0) {
        fprintf(stderr, "Failed to request button2 events\n");
        return -1;
    }
    
    return 0;
}

/*
 * Cleanup GPIO resources
 * Always important to release resources in embedded systems
 */
void gpio_cleanup(gpio_context_t *ctx) {
    if (ctx->led) gpiod_line_release(ctx->led);
    if (ctx->button1) gpiod_line_release(ctx->button1);
    if (ctx->button2) gpiod_line_release(ctx->button2);
    if (ctx->chip) gpiod_chip_close(ctx->chip);
}

/*
 * Handle button press event
 * This simulates an interrupt service routine (ISR)
 */
void handle_button_event(gpio_context_t *ctx, struct gpiod_line *line, 
                        const char *button_name, int *count) {
    struct gpiod_line_event event;
    
    // Read the event (clears the event flag)
    if (gpiod_line_event_read(line, &event) == 0) {
        (*count)++;  // Increment counter
        
        // Print event details
        printf("[%ld.%06ld] %s pressed! Count: %d\n", 
               event.ts.tv_sec, event.ts.tv_nsec / 1000,
               button_name, *count);
        
        // Toggle LED on button press
        static int led_state = 0;
        led_state = !led_state;  // Toggle between 0 and 1
        gpiod_line_set_value(ctx->led, led_state);
    }
}

/*
 * Main event loop using poll-style waiting
 * This is more efficient than continuous polling
 */
int main(void) {
    gpio_context_t gpio_ctx;
    struct gpiod_line_bulk bulk;  // For monitoring multiple lines
    struct timeval timeout;
    int ret;
    
    // Install signal handler
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);
    
    // Initialize GPIO
    if (gpio_init(&gpio_ctx) < 0) {
        fprintf(stderr, "GPIO initialization failed\n");
        return 1;
    }
    
    // Create a "bulk" object to monitor multiple GPIO lines
    // This allows efficient monitoring of multiple inputs
    gpiod_line_bulk_init(&bulk);
    gpiod_line_bulk_add(&bulk, gpio_ctx.button1);
    gpiod_line_bulk_add(&bulk, gpio_ctx.button2);
    
    printf("GPIO Event Detection Example\n");
    printf("Press buttons to trigger events. Ctrl+C to exit.\n");
    printf("Button 1: GPIO%d, Button 2: GPIO%d\n\n", BUTTON1_PIN, BUTTON2_PIN);
    
    // Main event loop
    while (running) {
        // Set timeout for event wait (1 second)
        // This allows periodic checking of the running flag
        timeout.tv_sec = 1;
        timeout.tv_usec = 0;
        
        // Wait for events on any monitored line
        // This is similar to interrupt-driven programming
        ret = gpiod_line_event_wait_bulk(&bulk, &timeout, NULL);
        
        if (ret < 0) {
            // Error occurred
            fprintf(stderr, "Event wait error: %s\n", strerror(errno));
            break;
        } else if (ret == 0) {
            // Timeout - no events
            // This is normal, just continue loop
            continue;
        }
        
        // Check which button(s) triggered events
        // Multiple events can occur simultaneously
        if (gpiod_line_event_get_fd(gpio_ctx.button1) >= 0) {
            // Check if this specific line has an event pending
            ret = gpiod_line_event_wait(gpio_ctx.button1, &timeout);
            if (ret > 0) {
                handle_button_event(&gpio_ctx, gpio_ctx.button1, 
                                  "Button 1", &gpio_ctx.button1_count);
            }
        }
        
        if (gpiod_line_event_get_fd(gpio_ctx.button2) >= 0) {
            ret = gpiod_line_event_wait(gpio_ctx.button2, &timeout);
            if (ret > 0) {
                handle_button_event(&gpio_ctx, gpio_ctx.button2, 
                                  "Button 2", &gpio_ctx.button2_count);
            }
        }
    }
    
    // Print statistics
    printf("\nFinal statistics:\n");
    printf("Button 1 pressed: %d times\n", gpio_ctx.button1_count);
    printf("Button 2 pressed: %d times\n", gpio_ctx.button2_count);
    
    // Cleanup
    gpio_cleanup(&gpio_ctx);
    
    return 0;
}

/*
 * Advanced C concepts demonstrated:
 * 1. typedef - Creates type aliases for easier code reading
 * 2. Structures with pointers - Organizing complex data
 * 3. Function pointers (hidden in gpiod library callbacks)
 * 4. Event-driven programming - Similar to interrupts
 * 5. Bulk operations - Efficient handling of multiple resources
 * 6. Proper error handling with errno
 * 
 * This example shows how Linux provides interrupt-like functionality
 * through file descriptors and event polling, which is different from
 * bare-metal embedded systems but achieves similar results.
 */

File: examples/02-gpio-interrupts/gpio-interrupts.py
#!/usr/bin/env python3
"""
Raspberry Pi 5 GPIO Interrupts - Event-Driven Programming

This example demonstrates event-driven GPIO programming using the gpiod library's
event detection capabilities. Instead of constantly polling, we wait for GPIO
state changes, which is more efficient.

Key concepts:
- Edge detection (rising/falling edges)
- Event timestamps
- Non-blocking I/O
- Multiple input monitoring

Hardware connections:
- Button 1: GPIO27 (pin 13) -> GND
- Button 2: GPIO22 (pin 15) -> GND  
- LED: GPIO17 (pin 11) -> 330Ω -> GND

Author: Your Name
Date: 2024
"""

import gpiod
import time
import signal
import sys
import select
from datetime import datetime

# GPIO Configuration
GPIO_CHIP = "gpiochip4"  # Pi 5's main GPIO chip
LED_PIN = 17            # GPIO17 for LED
BUTTON1_PIN = 27        # GPIO27 for button 1
BUTTON2_PIN = 22        # GPIO22 for button 2

# Event types
RISING_EDGE = gpiod.LINE_REQ_EV_RISING_EDGE
FALLING_EDGE = gpiod.LINE_REQ_EV_FALLING_EDGE
BOTH_EDGES = RISING_EDGE | FALLING_EDGE

# Global flags
running = True
press_count = 0
led_state = False

def signal_handler(sig, frame):
    """Handle Ctrl+C for graceful shutdown"""
    global running
    print("\nShutting down...")
    running = False

def format_timestamp(timestamp_ns):
    """Convert nanosecond timestamp to readable format"""
    # Convert nanoseconds to seconds
    timestamp_s = timestamp_ns / 1_000_000_000
    dt = datetime.fromtimestamp(timestamp_s)
    return dt.strftime("%H:%M:%S.%f")[:-3]  # Show milliseconds

def handle_button_event(event, button_num, led_line):
    """Process button events and control LED"""
    global press_count, led_state
    
    # Get event details
    timestamp = format_timestamp(event.timestamp)
    
    if event.type == RISING_EDGE:
        # Button released (due to pull-up)
        print(f"[{timestamp}] Button {button_num} released")
        
    elif event.type == FALLING_EDGE:
        # Button pressed
        press_count += 1
        print(f"[{timestamp}] Button {button_num} pressed! (Total: {press_count})")
        
        # Toggle LED
        led_state = not led_state
        led_line.set_value(1 if led_state else 0)
        print(f"LED: {'ON' if led_state else 'OFF'}")

def main():
    """Main program function"""
    print("=== Raspberry Pi 5 GPIO Interrupts Demo ===")
    print("Event-driven GPIO Programming\n")
    
    # Register signal handler
    signal.signal(signal.SIGINT, signal_handler)
    
    # Open GPIO chip
    try:
        chip = gpiod.Chip(GPIO_CHIP)
    except Exception as e:
        print(f"Failed to open GPIO chip: {e}")
        sys.exit(1)
    
    # Get GPIO lines
    try:
        led_line = chip.get_line(LED_PIN)
        button1_line = chip.get_line(BUTTON1_PIN)
        button2_line = chip.get_line(BUTTON2_PIN)
    except Exception as e:
        print(f"Failed to get GPIO lines: {e}")
        chip.close()
        sys.exit(1)
    
    # Configure LED as output
    try:
        led_line.request(consumer="gpio-interrupts",
                        type=gpiod.LINE_REQ_DIR_OUT,
                        default_vals=[0])
    except Exception as e:
        print(f"Failed to configure LED: {e}")
        chip.close()
        sys.exit(1)
    
    # Configure buttons for event detection
    # We request both edges to see press and release
    try:
        button1_line.request(consumer="gpio-interrupts",
                           type=gpiod.LINE_REQ_EV_BOTH_EDGES,
                           flags=gpiod.LINE_REQ_FLAG_BIAS_PULL_UP)
        
        button2_line.request(consumer="gpio-interrupts",
                           type=gpiod.LINE_REQ_EV_BOTH_EDGES,
                           flags=gpiod.LINE_REQ_FLAG_BIAS_PULL_UP)
    except Exception as e:
        print(f"Failed to configure buttons: {e}")
        led_line.release()
        chip.close()
        sys.exit(1)
    
    print("GPIO Configuration:")
    print(f"  LED: GPIO{LED_PIN}")
    print(f"  Button 1: GPIO{BUTTON1_PIN}")
    print(f"  Button 2: GPIO{BUTTON2_PIN}")
    print("\nWaiting for button events...")
    print("Press either button to toggle LED")
    print("Press Ctrl+C to exit\n")
    
    # Get file descriptors for event monitoring
    button1_fd = button1_line.event_get_fd()
    button2_fd = button2_line.event_get_fd()
    
    # Main event loop
    while running:
        # Use select to wait for events on either button
        # Timeout of 0.1s allows checking the running flag
        readable, _, _ = select.select([button1_fd, button2_fd], [], [], 0.1)
        
        # Check for button 1 events
        if button1_fd in readable:
            event = button1_line.event_read()
            if event:
                handle_button_event(event, 1, led_line)
        
        # Check for button 2 events
        if button2_fd in readable:
            event = button2_line.event_read()
            if event:
                handle_button_event(event, 2, led_line)
    
    # Cleanup
    print(f"\nTotal button presses: {press_count}")
    print("Cleaning up GPIO...")
    
    led_line.set_value(0)  # Turn off LED
    led_line.release()
    button1_line.release()
    button2_line.release()
    chip.close()
    
    print("Goodbye!")

if __name__ == "__main__":
    main()

"""
Advanced Concepts Explained:

1. Event-Driven vs Polling:
   - Polling: Constantly checking GPIO state (wastes CPU)
   - Event-Driven: Wait for state changes (efficient)
   - Uses select() to monitor multiple inputs

2. Edge Detection:
   - Rising Edge: LOW to HIGH transition
   - Falling Edge: HIGH to LOW transition
   - With pull-up: Press = falling, Release = rising

3. Event Timestamps:
   - Hardware timestamps for precise timing
   - Useful for debouncing and timing analysis

4. File Descriptors:
   - Linux treats everything as files
   - GPIO events use file descriptors
   - select() monitors multiple descriptors

5. Non-Blocking I/O:
   - Program can do other work while waiting
   - More responsive than polling
   - Better for battery-powered devices

Extending This Example:
- Add debouncing logic using timestamps
- Count press duration for long-press detection
- Monitor more inputs simultaneously
- Create interrupt service routines
"""